### ISSUE TITLE
Using Postgres as a database

### ISSUE THREAD

**Issue by thomaseizinger** [itchysats/itchysats#1545]
The goal of this ticket is to migrate the maker's db to a Postgres DB but keep the taker with SQLite. 

This includes a docker-based postgres setup for our local dev enivornment. 

Note: before merging this change we also need to upgrade our testnet setup and provide a postgres container

<details>
<summary>Old ticket details ⬇️</summary>
1. Do a PoC:

A rough outline of what is required to make this change. F.e. do we need to adjust the migrations or are they all standard conform. What about authentication? etc

The expected benefits of switching to postgres are:

- Improved performance
- Concurrent connections
- Move reporting out of the daemon into a separate program that queries the database

2. Based on PoC, decide on:

- Use postgres for maker and taker?
- If only for maker, how easy is it to create two database modules?
- Should we use sqlx's [`AnyConnection`](https://docs.rs/sqlx/latest/sqlx/struct.AnyConnection.html)?

## Annoyance counter

- 24/03/2022: Cannot easily access the production database and had to muck around with sshfs to get the sqlite file off the virtual machine.

</details>

**Comment by rishflab** [comment]
I think we can separate this into 2 main categories of issues, production related and sqlite migration to postgres related:

Sqlite to postgres migration:
- [x] If we want to maintain sqlite compatibility how do we ensure this?
- [x] Do our existing queries work
- [x] Do our existing migrations work?
- [ ] Do we want to migrate existing data from sqlite to postgres?

Production:
- [ ] How to deploy production postgres (same vm vs seperate vm vs managed)
- [ ] How to manage production postgres (debugging, logs, upgrades)
- [ ] Authentication/secrets handling

Integration with metrics:
- [ ] grafana

Developer experience
- [ ] postgres dev setup

**Comment by thomaseizinger** [comment]
This is a good start, thanks for writing this up!

The primary focus for now should be on figuring out, which code changes are required for switching to postgres.

**Comment by thomaseizinger** [comment]
PoC was posted here: https://github.com/itchysats/itchysats/pull/1631

**Comment by rishflab** [comment]
ah crap my commit containing the findings did not push. I just pushed it right now. Sorry!

**Comment by thomaseizinger** [comment]
With what I just worked on, I was using multiple git worktrees and thus now have different databases (because they are different directories).

In this case, it would be have nice to also connect the taker to a postgres DB that I can just run locally. Docker makes that very easy.

Our target userbase is umbrel which uses docker-compose and thus, running a postgres would be no problem there.
Running postgres for both parties would allow us to not split the DB for now which probably makes this feature quicker to implement.

Thoughts?

**Comment by bonomat** [comment]
If possible, I would try to keep the taker on SQLite. If we move them we will have to think about how to migrate all their data to the new DB. Doable, but not sure if we want to jump through that hoop. 

Also, we know that there are some users who are outside of Umbrel.

**Comment by thomaseizinger** [comment]
Yes, it obviously comes at a cost. But so does duplicating all the migration files and/or the database code. That is why I would like to approach this issue by putting all the facts on the table before we move forward with an implementation. To me it is not obvious what the best balance of trade-offs is.

**Comment by bonomat** [comment]
My notes to this topic: 


# Postgres vs SQLite

## Why?
### DevOps

- We want a managed db for our maker
- This allows us to start the maker container without any storage requirements as we can use a managed db (e.g. CloudSQL)
   - Could use kubernetes for simpler setup which would also allow us to do rolling upgrades-
 -  Performance
 	- We assume that Postgres gives us better performance
- Postgres gives us concurrent connections (we could access the database while in production)
  - We can connect to the DB using other services (e.g. grafana)

##  Why Not?
- SQLite is portable and platform independent. Users can just move the db file and start itchysats on a different platform. Postgres can't do it. For that the user needs to export the db into a file first
- SQLite does not need anything to be installed: no additional service or container.
    
## Research results

- We assume that Postgres is more performant than SQLite. According to "The Internet" this is not the case. (https://www.sqlite.org/speed.html): says that SQLite is 10-20 times faster than Postgres!
- Postgres needs more Ram which could be an issue with on RPi
- If performance was the driving factor: MySQL seems to perform better than Postgres. At some places you can even read that Postgres has performance issues (without evidences though: https://devathon.com/blog/mysql-vs-postgresql-vs-sqlite/) 

## Summary

Without verifying myself, it looks like that the performance assumption is wrong: SQLite seems to perform better for simple read operations. For more complex tasks, Postgres seems to be tool of choise.
For us, speed is important, which speaks against Postgres. Also, the hurdle to start an additional service is a negative point: at least for the end user.
If the main reason is performance, we should not switch away from Sqlite: the hurdle to change our migration, change the code to support postgres (or support sqlite and postgres) is too big.
This leaves us with the second reason: we want concurrent access to the db (for the maker) and move away from a file-based db so that our setup can be simplified.

Imho, we should look into how we could support Postgres for the maker and Sqlite for the taker

**Comment by thomaseizinger** [comment]
Thank you for the detailed summary!

Good to know that the performance assumption is debunked. Luckily we don't have to rely on that and all planned performance improvements should work with postgres as well.

Supporting different database technologies raises the question of whether or not to split the database into two source-code components, one per binary. That would lead to some duplication in the short-term (primarily caused by other shared components such as the process manager depending on the DB) but those could be mitigated through different means.

I see three paths forward:

1. Draft a vision for how the maker and taker binaries are meant to evolve. For example, do we see those being different binaries for a while or do we rather want to try and bring them closer together? Gaining confidence in either direction will help with forming a decision on whether or not the DB code can / should be split. For example, splitting the code now will substantially raise the efforts of trying to re-combine maker and taker again in the future.
2. Do a complete switch from sqlite to postgres, even for the taker. This creates some inconvenience for users that use itchysats outside of umbrel but has the big upside that we do not yet have to make a decision on the above question. We can defer that until later and unblock this issue.
3. Use the same database code with different databases. sqlx has an `AnyConnection` allows you to choose the database technology at runtime. This might allow us to defer the above question as well but will create operational complexity because we for example need to support both databases within our test suite. We also need to split at least the migrations as our current migrations are not compatible postgres (see https://github.com/itchysats/itchysats/pull/1631). It might also limit us in what features we can use within our SQL code.

Personally, I have a mild preference for option (2) over (3) over (1):

- I think a vision on where `maker` and `taker` are heading needs to be formed at some point but there is no need to block this issue on it.
- (2) should be simpler to implement and maintain than (3) for us.
- The downside of (2) is only an inconvenience for some users that don't run inside umbrel.
- Those users are a minority AFAIK.
- It is still possible to use it outside of umbrel.
- Users that run outside of the recommended parameters already accept a certain level of "tinkering around".

**Comment by thomaseizinger** [comment]
In a meeting, we recently chatted about the vision of maker and taker binaries previously. I am yet to draft something more concrete up but my proposal would be the following:

- Let maker and taker binaries diverge so we can customize them to our current needs.
- Push for consolidation of "shared" code into independent crates. This will require changes to internal interfaces but it should be possible to carve out re-usable modules that are role-independent.

Before we can even start thinking about a combined binary for maker and taker, I believe we would first need to come up with network protocols that make sense to be triggered by either party (like placing orders). Before that, it seems to me that letting them diverge gives us more flexibility.

For the issue at hand, this would mean we could start duplicating the DB code and modify them independently. I am worried about the temporary code duplication though so even in light of this, my preference would still go for (2).

**Comment by thomaseizinger** [comment]
With the recent introduction of the `db::Connection` abstraction, it should be easier to duplicate only the DB code and have other shared components such as the process manager still shared. Hence, moving only the maker to postgres and leaving the taker on sqlite is easier now. (Changes like this are what I meant with "but those could be mitigated through different means." above).

One concrete option would be to lift a trait out of `db::Connection` and implement it for two particular ones, one for sqlite and one for postgres. Components like the process manager can then depend on a trait object (`Box<dyn db::Connection>`).

With this change, the size of the `daemon` crate will grow once again. It would be wise to eventually push https://github.com/itchysats/itchysats/pull/1337 forward to make it possible for newly introduced module like the postgres DB and the sqlite DB to live in the respective crates where they are actually used (or perhaps in separate crates altogether). The current `db` module looks like it is quite standalone so it may be worth trying to split this out into a separate crate.

**Comment by thomaseizinger** [comment]
> The current `db` module looks like it is quite standalone so it may be worth trying to split this out into a separate crate.

A quick PoC shows that it is actually possible: https://github.com/itchysats/itchysats/commit/1b2f851849697f30cd409460c57a468752b726ae

**Comment by luckysori** [comment]
> > The current `db` module looks like it is quite standalone so it may be worth trying to split this out into a separate crate.
> 
> A quick PoC shows that it is actually possible: [1b2f851](https://github.com/itchysats/itchysats/commit/1b2f851849697f30cd409460c57a468752b726ae)

If we go for this we will need to move some of the db tests so that we don't introduce a circular dependency:

https://github.com/itchysats/itchysats/blob/1b2f851849697f30cd409460c57a468752b726ae/sqlite-db/src/lib.rs#L1734

**Comment by bonomat** [comment]
A goal of this week is to come to a conclusion on this topic: 

given the latest findings from @thomaseizinger , do we agree to go for this?

1. taker stays with SQLite
2. maker migrates to Postgres - 
    1. for our local dev setup: if "easily" possible, it would be nice to support both dbs for the maker, i.e. SQLite and Postgres. If it is deemed to be too complex or too much work, then a simple docker-based setup for devs is enough.

**Comment by bonomat** [comment]
Once we have https://github.com/itchysats/itchysats/issues/2139 we should not have any sqlx types lying around anywhere else besides in the db module. We should then be able to tackle this ticket by introducing a new module which deals with postgres.

**Comment by bonomat** [comment]
This project is unmaintained
